### DISCUSSION TITLE
EvalEx 3.0.0 released

### DISCUSSION THREAD

**Discussion by uklimaschewski** [ezylang/EvalEx#313] [Announcements]
EvalEx 3 has been released and is also available at Maven central!

I am very happy about this release and also about the new version 3.
Thank you to all, who supported EvalEx!

Please notice the new Maven coordinates for your dependency management:

```xml
<dependency>
  <groupId>com.ezylang</groupId>
  <artifactId>EvalEx</artifactId>
  <version>3.0.0</version>
</dependency>
```

Though the expressions used with EvalEx 2 should also evaluate on version 3 correctly, there are several breaking changes in the API.
See https://ezylang.github.io/EvalEx/concepts/changes.html for an overview of the major changes.

EvalEx 3 comes with a lot of new features, you can explore them in the new documentation web site at https://ezylang.github.io/EvalEx/index.html

This is the first release of EvalEx 3. If you find any bugs, file an issue here: https://github.com/ezylang/EvalEx/issues
Pull requests and other contributions with fixes and extensions are welcome!

For discussions, questions, ideas and announcements, visit the discussion area here: https://github.com/ezylang/EvalEx/discussions

**Comment by AlexandrKozlovskiy** [comment]
@uklimaschewski It's very good,what you released new version of library. I not see sources yet,so i have a few questions.

1. In evalex 2,as you wrote,i can set in loop,for example,from 1 to 10 values of some variables,and it not parsed again. Whether it works the same,or evalex reparse expression for each iteration,so it will be problems with performance in loop. Can i reuse the same object of expression,because when we set a new values of variables,method returns a new instance of exprrssion object.
2. Whether exists way to catch,if memory overflow was happen,because it can be for very long values.
3. Whether ast more quickly,then rpm?or it more slowely,so we can have problems with performance?
4. Whether we can override existing function and operators,and can we,for example,use expression x++,or x+=1,what will mean expression x+1. The same i want to have for minus operator.

**Reply by uklimaschewski** [reply]
The sources are here :-)
1) Yes, the "compiled" expression is cached just like before and can be executed with different values multiple times.
2) You can, but it makes no real sense to catch an OutOfMemory error in java, because of the Java memory design, you can't predict where it will happen.
3) I did no real performance comparison. EvalEx was never designed to do high performance calculations.
4) Overriding functions and operators is possible and even simpler than before. Above operators are possible.

have a look at the new documentation at https://ezylang.github.io/EvalEx.

**Reply by AlexandrKozlovskiy** [reply]
@uklimaschewski Thank you very much for your reply. It very good,what expression parsed once,so,configuration,which i set,saves,when i use with or and method of expression class,so methods with and and return the same instance of expression object. Can you please add information to readme,how i can connect evalex via gradle,because it seems,what for evalex 2
 i could use it with gradle in my android calculator app.

**Reply by uklimaschewski** [reply]
Of course you can use EvaEx 3 also with any build system, also with gradle.
Best way to find out how to do it, is to look at the Maven Central page:
https://search.maven.org/artifact/com.ezylang/EvalEx/3.0.0/jar
It lists templates for all build systems, ready to copy and paste.

**Reply by AlexandrKozlovskiy** [reply]
@uklimaschewski Thank you very much. Can you please add this information to readme file too?

**Reply by uklimaschewski** [reply]
Thank you, I added that to the README.
